datafusion/optimizer/src/analyzer/rewrite_expr.rs

```diff
@@ -154,103 +154,35 @@ fn rewrite_array_concat_operator_to_func(
         return None;
     }
 
-    match (left, right) {
+    fn fn_name(expr: &Expr) -> Option<&str> {
+        if let Expr::ScalarFunction(ScalarFunction { func_def, .. }) = expr {
+            Some(func_def.name())
+        } else {
+            None
+        }
+    }
+
+    // TODO figure out how to generalize this so it isn't hard coded by name
+    match (fn_name(left), fn_name(right)) {
```

Comment:
I am not sure why we have specific rewrites here for functions (that can't be done by constant folding or some other method). I don't like special casing the names

I need to look into it